covid19-dashboard/main.py

```diff
@@ -12,201 +12,391 @@
-                                        compress=True)
-    county_response: dict = send_request(DC_SERVER + "bulk/place-obs",
-                                         dict(placeType='County',
-                                              populationType="Person",
-                                              observationDate="2018"),
-                                         api_key=API_KEY,
-                                         compress=True)
-
-    response: list = state_response['places'] + county_response['places']
-
-    output: defaultdict = defaultdict(int)
+def state_data():
+    """
+    Returns US state data. Faster than /county-data api.
+    NOTE: for return type documentation, please see README.md's APIs section.
+    :return: geo_id->{**data, name, geo_id, regionType, containedIn}.
+    """
+    # Request data for only US States.
+    data = get_data("State")
+    # Adds HTTP headers for browser to store cache.
+    response = add_browser_cache_headers_to_response(data)
+    return response
+
+
+def get_data(place_type: str = "State") -> OutputDataType:
+    """
+    Requests cases and deaths from the DC, and performs calculations.
+    NOTE: for return type documentation, please see README.md's APIs section.
+    :param place_type: return states OR counties.
+    :return: geo_id->{**data, name, geo_id, regionType, containedIn}.
+    """
+    # Get all the places in the US.
+    # States and Counties.
+    places: RegionToInfoType = get_us_places(place_type)
+    # Get the total population of all geo_id.
+    population: Dict[str, int] = get_population()
+
+    # Convert the places to a list of geo_id.
+    # The dictionary is keyed by geo_id, so we can just use that.
+    geo_ids: List[str] = list(places.keys())
+
+    # Request all cases and deaths.
+    all_cases: GeoIdToDataType = get_stats_by_date(geo_ids,
+                                                   INCREMENTAL_CASES_STATVAR)
+    all_deaths: GeoIdToDataType = get_stats_by_date(geo_ids,
+                                                    INCREMENTAL_DEATHS_STATVAR)
+
+    output: OutputDataType = {}
+
+    for geo_id in geo_ids:
+        calculated_cases: Dict[str, int] = {}
+        calculated_deaths: Dict[str, int] = {}
+        # Make sure we have the population for this geo_id.
+        if geo_id not in population:
+            continue
+        # Otherwise, get the population for this geo_id.
+        region_pop = population[geo_id]
+        # Get the data for this geo_id.
+        if geo_id in all_cases:
+            cases = all_cases[geo_id]
+            # Perform XYZ calculations.
+            calculated_cases = calculate_data("Cases", cases, region_pop)
+        if geo_id in all_deaths:
+            deaths = all_deaths[geo_id]
+            # Perform XYZ calculations.
+            calculated_deaths = calculate_data("Deaths", deaths, region_pop)
+
+        name: str = ""
+        contained_in_geo_id: str = ""
+        place_type: str = ""
```

Comment:
It's generally not good practice to override a function parameter like this.

Reply:
No need to overwrite it  since this function only gets data based on place_type input parameter, so all should be of the same place_type. Not necessary to retrieve it again.

Reply:
Removed overwrite. Thanks